### DISCUSSION TITLE
Release Schedule

### DISCUSSION THREAD

**Discussion by cdnninja** [audiconnect/audi_connect_ha#272] [General]
I have enabled releases as a starting point.  Feedback is appreciated on if it should be changed.  

Right now it auto generates release notes and increases numbers via semantic version control.  It releases on Wed/Sunday.  Any owner can also trigger it via actions.    I am open to adjusting it if it doesn't make sense.  This lets us also have uses test things in master before it hits widespread updates.

**Comment by cdnninja** [comment]
In case it interests others here is the current stats for installs for those with the reporting turned on.   With the recent releases our version will take time to catch up.   Some of those versions are also other forks so won't update until the person reverts to this one. 

    "audiconnect": {
        "total": 648,
        "versions": {
            "1.0.1": 50,
            "1.0.0": 18,
            "2021.1.1": 290,
            "1.3.0": 126,
            "1.0.3": 29,
            "1.2.0": 33,
            "1.1.0": 59,
            "1.0.2": 25,
            "1.5.1": 5,
            "2023.12.3": 11,
            "1.0": 2
        }
    },

Source: https://[analytics.home-assistant.io/custom_integrations.json](https://analytics.home-assistant.io/custom_integrations.json)

**Reply by Kolbi** [reply]
```
"audiconnect":{
      "total":660,
      "versions":{
         "1.0.1":41,
         "1.0.0":14,
         "2021.1.1":290,
         "1.3.0":67,
         "1.5.0":33,
         "1.3.1":36,
         "1.5.1":14,
         "1.4.0":4,
         "1.2.0":13,
         "1.5.2":65,
         "1.1.0":32,
         "1.0.2":18,
         "1.0.3":19,
         "2023.12.3":12,
         "1.0":2
      }
   }
```

Wow high adapting rate currently :)

BTW: sorting is interesting in the json :)

**Reply by coreywillwhat** [reply]
```json
{
   "audiconnect": {
      "total": 685,
      "versions": {
         "1.0.1": 33,
         "1.0.0": 14,
         "2021.1.1": 256,
         "1.6.5": 75,
         "1.6.4": 141,
         "1.5.2": 14,
         "1.5.3": 23,
         "1.5.0": 4,
         "1.1.0": 14,
         "1.0.2": 15,
         "1.3.0": 17,
         "1.2.0": 5,
         "1.5.1": 3,
         "1.3.1": 9,
         "1.4.0": 1,
         "1.6.2": 12,
         "1.6.1": 5,
         "2023.12.3": 9,
         "1.0": 2,
         "1.6.3": 14,
         "1.6.0": 6,
         "1.5.4": 3,
         "1.7.0": 7,
         "1.0.3": 3
      }
   }
}
```

**Reply by cdnninja** [reply]
Amazing work!!

**Reply by Kolbi** [reply]
Still growing:

>   "audiconnect": {
>     "total": 705,
>     "versions": {
>       "1.0.1": 26,
>       "1.0.0": 13,
>       "2021.1.1": 235,
>       "1.8.0": 113,
>       "1.9.0": 99,
>       "1.5.3": 8,
>       "1.5.0": 4,
>       "1.5.2": 5,
>       "1.7.0": 65,
>       "1.6.4": 43,
>       "1.0.2": 9,
>       "1.6.5": 15,
>       "1.5.1": 1,
>       "1.3.0": 13,
>       "1.1.0": 10,
>       "1.3.1": 7,
>       "1.6.1": 1,
>       "2023.12.3": 9,
>       "1.0": 2,
>       "1.6.2": 6,
>       "1.6.3": 7,
>       "1.5.4": 6,
>       "1.2.0": 4,
>       "1.6.0": 2,
>       "1.0.3": 2
>     }
>   },

**Reply by cdnninja** [reply]
This is also only users who enroll in the telemetry. So actual user count is higher.

**Reply by Kolbi** [reply]
Just last two versions:
      "1.9.0": 367,
      "1.10.0": 140,

"1.10.1" not yet mentioned.

**Reply by Kolbi** [reply]
Just to update this list:

>   "audiconnect": {
    "total": 1176,
    "versions": {
      "1.10.4": 1071,
      "1.10.1": 8,
      "1.10.2": 34,
      "2021.1.1": 32,
      "1.10.3": 8,
      "1.7.0": 3,
      "1.9.0": 10,
      "1.3.0": 1,
      "1.0": 1,
      "1.8.0": 1,
      "2023.12.3": 1,
      "1.0.1": 2,
      "2.1.82": 1,
      "1.0.3": 1,
      "1.10.0": 2
    }

**Reply by Kolbi** [reply]
>   "audiconnect": {
    "total": 1180,
    "versions": {
      "1.10.4": 704,
      "1.11.0": 344,
      "1.10.1": 8,
      "1.10.2": 31,
      "2021.1.1": 32,
      "1.10.3": 6,
      "1.7.0": 3,
      "1.9.0": 10,
      "1.3.0": 1,
      "1.0": 1,
      "1.8.0": 1,
      "2023.12.3": 1,
      "1.11.1": 32,
      "1.0.1": 2,
      "2.1.82": 1,
      "1.0.3": 1,
      "1.10.0": 2
    }

**Reply by Kolbi** [reply]
"audiconnect": {
    "total": 1187,
    "versions": {
      **"1.12.0": 425,**
      "1.11.1": 63,
      "1.10.4": 314,
      "1.11.2": 252,
      "1.10.1": 7,
      "1.10.2": 25,
      "1.11.0": 10,
      "2021.1.1": 28,
      "1.7.0": 3,
      "1.9.0": 9,
      "1.10.3": 3,
      "1.3.0": 1,
      "1.0": 1,
      "1.8.0": 1,
      "2023.12.3": 1,
      **"1.12.1": 40,**
      "1.0.1": 2,
      "1.0.3": 1,
      "1.10.0": 1
    }

**Reply by Kolbi** [reply]
"audiconnect": {
    "total": 1421,
    "versions": {
      "1.16.1": 724,
      "1.14.0": 245,
      "1.10.4": 141,
      "1.13.2": 28,
      "1.16.0": 118,
      "1.11.2": 3,
      "1.15.0": 71,
      "1.10.2": 15,
      "2021.1.1": 13,
      "1.10.1": 3,
      "1.12.1": 4,
      "1.13.0": 6,
      "1.13.1": 16,
      "1.10.3": 2,
      "1.3.0": 1,
      "1.0": 1,
      "1.9.0": 3,
      "1.12.2": 1,
      "1.12.0": 3,
      "1.0.1": 1,
      "1.17.0": 20,
      "1.12.4": 1,
      "1.7.0": 1
    }
  },

**Reply by Kolbi** [reply]
"audiconnect": {
    "total": 1683,
    "versions": {
      "1.17.3": 1413,
      "1.17.0": 55,
      "1.10.4": 111,
      "1.13.2": 8,
      "1.14.0": 24,
      "1.16.0": 3,
      "1.10.2": 13,
      "1.17.2": 14,
      "2021.1.1": 9,
      "1.10.1": 5,
      "1.12.1": 2,
      "1.12.4": 2,
      "1.13.1": 6,
      "1.10.3": 2,
      "1.9.0": 3,
      "1.3.0": 1,
      "1.15.0": 7,
      "1.0": 1,
      "1.12.0": 2,
      "1.0.1": 1,
      "1.16.1": 1
    }
  },

**Reply by Kolbi** [reply]
"audiconnect": {
    "total": 2003,
    "versions": {
      "1.19.1": 1627,
      "1.17.0": 17,
      "1.13.2": 3,
      "1.10.4": 93,
      "1.18.1": 44,
      "1.19.0": 98,
      "1.14.0": 11,
      "1.17.3": 37,
      "1.16.0": 3,
      "1.10.2": 11,
      "2.0.1-beta.9": 23,
      "1.17.2": 3,
      "2021.1.1": 9,
      "1.10.1": 3,
      "1.12.1": 1,
      "1.13.1": 3,
      "1.10.3": 2,
      "1.3.0": 1,
      "1.0": 1,
      "1.9.0": 2,
      "2.0.1-beta.8": 4,
      "1.12.0": 1,
      "1.15.0": 3,
      "1.18.0": 1,
      "1.12.4": 1,
      "1.0.1": 1
    }

**Reply by Kolbi** [reply]
"audiconnect": {
    "total": 1783,
    "versions": {
      "2.2.0": 538,
      "2.3.0": 718,
      "1.17.0": 7,
      "1.10.4": 50,
      "2.1.2": 44,
      "1.13.2": 1,
      "1.19.1": 95,
      "2.1.4": 125,
      "1.19.0": 16,
      "2.2.1b7": 9,
      "1.14.0": 7,
      "2.2.1b8": 20,
      "2.2.1b1": 39,
      "2021.1.1": 7,
      "2.2.1b5": 15,
      "1.18.1": 10,
      "2.2.1b6": 1,
      "2.3.1b1": 8,
      "1.17.3": 13,
      "1.10.1": 2,
      "1.13.1": 3,
      "1.19.2": 8,
      "2.2.1b2": 8,
      "2.1.1": 6,
      "1.10.2": 7,
      "1.3.0": 1,
      "2.0.1-beta.11": 1,
      "1.0": 1,
      "1.9.0": 2,
      "1.16.0": 1,
      "2.1.3": 8,
      "2.2.1b3": 3,
      "2.0.1-beta.9": 1,
      "1.10.3": 1,
      "2.1.0": 1,
      "1.0.1": 1,
      "1.15.0": 1,
      "1.19.0+idk-fix2": 1,
      "2.3.1b2": 2,
      "2.2.1b4": 1
    }

**Comment by coreywillwhat** [comment]
Is there a way we can run a beta branch? There seems to be a lot of variability with vehicles. Some reporting different timestamp formats, or different features supported or not. it would be nice to push some beta releases so those who want to test can, and others can stay on the stable version.

**Reply by cdnninja** [reply]
We can set releases to beta by hand. I'm sure can be done with actions as well. 

Are you looking for auto beta releases or manual?

**Reply by coreywillwhat** [reply]
I'd be fine to do the beta releases manually.
If we're doing that, we would probably want to consider pausing the auto releases for stable, as we wouldn't want stable releases pushed with beta code until its stable? 

Is it possible to automatically push beta releases, and manually push stable? Or would we create a new branch to run the beta code and just push from there? Haven't done this before so curious what the right path is for this kind of scenario.

@Kolbi was the unfortunate victim of code that was untested on a vehicle with no sunroof.
The other solution might be buying a bunch of audis so we can appropriately test :)

**Reply by cdnninja** [reply]
Right now it generates a release Sunday and Wednesday.  

If some merges need to be beta you can run the action to generate the release and right after edit the release and select the beta checkbox.   

The action will release a stable release next commit though on those days.   

We could make the release action always do beta releases but someone needs to check the box to convert to non beta at some point.  Manually is always a challenge since this is a group of people it may get forgotten.   

Part of the two days a week approach was it lets people update from master to test if needed. However that's also manual.  

Open to thoughts.

**Reply by coreywillwhat** [reply]
I'm torn between going full auto release for beta with the thought being, if someone is in here submitted PRs and merging, then they will likely know to either push a manual or wait for a beta push.

The other part of me is thinking move the stable releases to weekly, to give more time for beta testing throughout the week?

My major concern here is I dont think there is enough community involvement to get people to manually patch from the master branch for testing. I do think there are people who would choose the beta branch within HACS/HA and update to that/feel comfortable with that. As they can always downgrade from within HA too.

**Reply by cdnninja** [reply]
If I find a way to say:

Release as beta on commit merge or once a day. 

Weekly promote the current beta to prod


It would mean if someone finds a breaking beta near end of week they need to pull that beta by hand. 

Otherwise would be automated.   Would that help?

**Reply by cdnninja** [reply]
Other note to keep in mind is HACS only checks in once every few days.  So the average person isn't notified of releases for a bit.  Most likely why that issue was caught by a dev vs general user.

**Reply by coreywillwhat** [reply]
That method makes sense to me, but no rush on this either. Would rather take the time we need to discuss and find something we're all comfortable with. Do you think the approach makes sense? I don't want to change something just because its something I'm thinking. Really appreciate the collaboration.

**Reply by Kolbi** [reply]
https://github.com/audiconnect/audi_connect_ha/pull/539#issuecomment-2736450989
Question is, do we want to release beta pre-release versions from the main branch? That should be an easy task, or do we want to create a dev branch with daily beta pre-releases? Also doable but I'm if we should create automated releases from all branches....

**Reply by cdnninja** [reply]
I don't have audi connect anymore which is why haven't spent much time on it.      I think it really depends on how many PRs go into this.   

In my experience force a release and tag it with pre-release for risky ones works well.  Its a quick checkmark.   However automated approach is good too just means more branch management.

**Reply by Kolbi** [reply]
Had some issue with next version because 1.11.3 was used already for the pre-release :)

We should use steps between for manual pre-release. For example last stable: 1.11.3 => pre-release 1.11.3.1

**Reply by coreywillwhat** [reply]
Ah okay, sounds good! I can do that! Getting really granular haha

**Reply by coreywillwhat** [reply]
For some reason 1.12.3.1 is showing before 1.12.3, I supposed because 1.12.3 would have to be 1.12.3.0 to think it should be before the pre-release.

**Reply by coreywillwhat** [reply]
I've been away from active development for a while, but wanted to reopen the conversation about setting up a beta/pre-release workflow.

I've been working on my own fork using Claude/Codex to do the heavy lifting while I review and test. I don't have the time anymore (kids) to code or put in the effort required to restructure things on my own. There are a lot of moving pieces involved. My proposal is to maintain a dedicated `beta` branch and publish pre-releases manually once I've reviewed and tested.

### Goal: v2.0

The goal of this beta would be to target v2.0, which involves significant restructuring and migration work. This way, we can:

- Keep **v1 stable** with normal `latest` release bug fixes for existing users
- Develop and test the **v2.0 major changes** in parallel without disrupting anyone on v1

We can also add guidance to the `README` for users who want to opt into the pre-release, so it's clear how to do so and what to expect.

This gives us a clean separation between ongoing maintenance and the bigger modernization effort. Let me know your guys thoughts.

edit: Other option is we go with a new repo called "myAudi ha" to align with the updated audi app name which we're using on the backend anyway, and eventually we leave this one behind.

edit 2: @Kolbi I'm putting all my effort into the beta branch at the moment. There are so many changes, i will try to keep up whenever changes are made to the master branch, but would love your help testing the beta branch as well as your input on pre-release.

**Reply by Kolbi** [reply]
Hi, releasing separate beta packages would definitly make sense in order to test it easier. 
I will try to catch up and install the beta. :)

**Reply by coreywillwhat** [reply]
> Hi, releasing separate beta packages would definitly make sense in order to test it easier. 
> 
> I will try to catch up and install the beta. :)

My fork has the latest if you want to install the latest prerelease from there. You can also see all the updates from each release there. That said, I've been at the hospital for a week so I am not sure where things stand and if I left any bugs unfixed. Let me know what you find!Thanks @Kolbi

**Reply by cdnninja** [reply]
I don't pay for services so don't use this anymore. Focused on a few other projects. 

As a side note also getting a potential engine replacement for the q7....

**Reply by coreywillwhat** [reply]
> As a side note also getting a potential engine replacement for the q7....

oh man good luck. Glad to hear the q7 will live on!

**Reply by coreywillwhat** [reply]
> Hi, releasing separate beta packages would definitly make sense in order to test it easier. I will try to catch up and install the beta. :)

@Kolbi I pushed the first pre-release if you want to test

**Reply by Kolbi** [reply]
@coreywillwhat I'm currently using latest beta. HACS showed me also the pre-release. I think by default prerelease are shown to endusers.
